### PR TITLE
fix: use proper lifetimes for cache and dataset

### PR DESCRIPTION
**Pull request by therustmonk**
The PR wraps pointers to a cache and a dataset with `Arc` that let to track the lifetime of the objects properly.

### CODE REVIEW

**Comment by therustmonk** [comment]
And the library is not thread-safe, we shouldn't implement `Sync` for that: https://github.com/tevador/RandomX/issues/215
Or we should wrap it in the `Mutex`.

That issue above is about a thread-safety of VM, but I'm not sure the cache and the dataset are thread-safe as well since they are used by a single VM. But we still can implement `Send` if those types don't use thread-local storage.

File: src/lib.rs [hunk 1]
```diff
@@ -296,104 +313,93 @@ impl RandomXVM {
         cache: Option<&RandomXCache>,
```
Comment (stringhandler) [inline src/lib.rs]: nit: I like to move objects when they will be cloned in the method
```suggestion
        cache: Option<RandomXCache>,
        dataset: Option<RandomXDataset>
```

File: src/lib.rs [hunk 2]
```diff
@@ -296,104 +313,93 @@ impl RandomXVM {
...
-        let result = RandomXVM { vm: test, flags };
-        Ok(result)
     }
 
     /// Re-initializes the `VM` with a new cache that was initialised without
     /// RandomXFlag::FLAG_FULL_MEM.
-    pub fn reinit_cache(&self, cache: &RandomXCache) -> Result<(), RandomXError> {
-        if self.flags & RandomXFlag::FLAG_FULL_MEM == RandomXFlag::FLAG_FULL_MEM {
-            return Err(RandomXError::FlagConfigError(
+    pub fn reinit_cache(&mut self, cache: &RandomXCache) -> Result<(), RandomXError> {
```
Comment (stringhandler) [inline src/lib.rs]: Same here
```suggestion
    pub fn reinit_cache(&mut self, cache: RandomXCache) -> Result<(), RandomXError> {
```

File: src/lib.rs [hunk 3]
```diff
@@ -296,104 +313,93 @@ impl RandomXVM {
...
         }
-        Ok(())
     }
 
     /// Re-initializes the `VM` with a new dataset that was initialised with
     /// RandomXFlag::FLAG_FULL_MEM.
-    pub fn reinit_dataset(&self, dataset: &RandomXDataset) -> Result<(), RandomXError> {
-        if self.flags & RandomXFlag::FLAG_FULL_MEM != RandomXFlag::FLAG_FULL_MEM {
-            return Err(RandomXError::FlagConfigError(
+    pub fn reinit_dataset(&mut self, dataset: &RandomXDataset) -> Result<(), RandomXError> {
```
Comment (stringhandler) [inline src/lib.rs]: Same here
```suggestion
    pub fn reinit_dataset(&mut self, dataset: RandomXDataset) -> Result<(), RandomXError> {
```

**Review by stringhandler — DISMISSED** [review]
Some nits, but otherwise perfect!

**Review by stringhandler — APPROVED** [review]

**Pull request merged**